src/java_bytecode/generate_java_generic_type.cpp

```diff
@@ -0,0 +1,102 @@
+/*******************************************************************\
+
+ Module: Generate Java Generic Type - Instantiate a generic class with
+         concrete type information.
+
+ Author: DiffBlue Limited. All rights reserved.
+
+\*******************************************************************/
+#include "generate_java_generic_type.h"
+#include <util/namespace.h>
+#include <java_bytecode/java_types.h>
+#include <java_bytecode/java_utils.h>
+
+#include <iostream>
+#include <sstream>
+#include <algorithm>
+
+generate_java_generic_typet::generate_java_generic_typet(
+  message_handlert &message_handler):
+    message_handler(message_handler)
+{}
+
+/// Generate a concrete instantiation of a generic type.
+/// \param existing_generic_type The type to be concretised
+/// \param symbol_table The symbol table that the concrete type will be inserted into.
+/// \return The symbol as it was retrieved from the symbol table after it has been inserted into.
+symbolt generate_java_generic_typet::operator()(
+  const java_generic_typet &existing_generic_type,
+  symbol_tablet &symbol_table) const
+{
+  PRECONDITION(existing_generic_type.id()==ID_pointer);
+  namespacet ns(symbol_table);
+
+  const typet &pointer_subtype=ns.follow(existing_generic_type.subtype());
+
+  INVARIANT(
+    pointer_subtype.id()==ID_struct, "Only pointers to classes in java");
+
+  const java_class_typet &replacement_type=to_java_class_type(pointer_subtype);
+  const irep_idt new_tag=build_generic_tag(existing_generic_type, replacement_type);
+  struct_union_typet::componentst replacement_components=replacement_type.components();
+
+  // Small auxiliary function, to perform the inplace modification of the generic fields.
+  auto replace_type_for_generic_field=
+    [&](struct_union_typet::componentt &component)
+    {
+      if(is_java_generic_parameter(component.type()))
+      {
+        INVARIANT(
+          existing_generic_type.generic_type_variables().size()==1,
+          "Must have a type parameter");
+        component.type() = existing_generic_type.generic_type_variables()[0];
+      }
+      return component;
+    };
+
+  std::size_t pre_modification_size=replacement_type.components().size();
+
+  std::for_each(
+    replacement_components.begin(),
+    replacement_components.end(),
+    replace_type_for_generic_field);
+
+  std::size_t after_modification_size=replacement_type.components().size();
+
+  INVARIANT(
+    pre_modification_size==after_modification_size,
+    "All components in the original class should be in the new class");
+
+  generate_class_stub(new_tag, symbol_table, message_handler, replacement_components);
+  INVARIANT(symbol_table.has_symbol("java::"+id2string(new_tag)), "New class not created");
+  return symbol_table.lookup("java::"+id2string(new_tag));
+}
+
+/// Build a unique tag for the generic to be instantiated.
```

Comment:
please give an example of how the tag will look like. Also it seems that you concat using `, `, I am not too fond of using spaces here

Reply:
This comment doesn't seem to have been addressed?